cmd/fs-v1-multipart.go

```diff
@@ -799,6 +800,19 @@ func (fs fsObjects) CompleteMultipartUpload(bucket string, object string, upload
 				PartETag:   part.ETag,
 			})
 		}
+		if partSize == -1 {
+			partSize = fsMeta.Parts[partIdx].Size
+		}
+		// All parts except the last part has to be of same size. We are introducing this
+		// check to see if any clients break. If clients do not break then we can optimize
+		// multipart PutObjectPart by writing the part at the right offset using pwrite()
+		// so that we don't need to do background append at all. i.e by the time we get
+		// CompleteMultipartUpload we already have the full file available which can be
+		// renamed to the main name-space.
+		if (i < len(parts)-1) && partSize != fsMeta.Parts[partIdx].Size {
```

Comment:
Should this check be done on upload multipart, instead of waiting till complete multipart?

Reply:
I tried to do this, but it affected test cases, fixing which was tricky and needs more effort.

Hence we can do this check in PutObjectPart once we know that no clients break and we incorporate the pwrite() style PutObjectPart. During which we can invest time in fixing the test cases as well. We have to fix test cases carefully as we allow parts with different sizes for erasure backend.